beluga_amcl/test/test_ndt_amcl_node.cpp

```diff
@@ -0,0 +1,734 @@
+
+    auto transform_laser = geometry_msgs::msg::TransformStamped{};
+    transform_laser.header.stamp = timestamp;
+    transform_laser.header.frame_id = "base_footprint";
+    transform_laser.child_frame_id = "laser";
+
+    laser_scan_publisher_->publish(scan);
+    tf_broadcaster_->sendTransform(transform_base);
+    tf_broadcaster_->sendTransform(transform_laser);
+  }
+
+  bool can_transform(const std::string& target, const std::string& source) const {
+    return tf_buffer_ && tf_buffer_->canTransform(target, source, tf2::TimePointZero);
+  }
+
+  auto lookup_transform(const std::string& target, const std::string& source) const {
+    auto transform = Sophus::SE2d{};
+    if (tf_buffer_) {
+      tf2::convert(tf_buffer_->lookupTransform(target, source, tf2::TimePointZero).transform, transform);
+    }
+    return transform;
+  }
+
+ private:
+  template <class Message>
+  using PublisherPtr = std::shared_ptr<rclcpp::Publisher<Message>>;
+
+  PublisherPtr<geometry_msgs::msg::PoseWithCovarianceStamped> initial_pose_publisher_;
+  PublisherPtr<sensor_msgs::msg::LaserScan> laser_scan_publisher_;
+
+  template <class Message>
+  using SubscriberPtr = std::shared_ptr<rclcpp::Subscription<Message>>;
+
+  SubscriberPtr<geometry_msgs::msg::PoseWithCovarianceStamped> pose_subscriber_;
+  SubscriberPtr<nav2_msgs::msg::ParticleCloud> particle_cloud_subscriber_;
+
+  std::optional<geometry_msgs::msg::PoseWithCovarianceStamped> latest_pose_;
+  std::optional<nav2_msgs::msg::ParticleCloud> latest_particle_cloud_;
+
+  std::unique_ptr<tf2_ros::Buffer> tf_buffer_;
+  std::unique_ptr<tf2_ros::TransformBroadcaster> tf_broadcaster_;
+  std::unique_ptr<tf2_ros::TransformListener> tf_listener_;
+};
+
+/// Base node fixture class with common utilities.
+template <class T>
+class BaseNodeFixture : public T {
+ public:
+  void SetUp() override {
+    rclcpp::init(0, nullptr);
+    ndt_amcl_node_ = std::make_shared<AmclNodeUnderTest>();
+    tester_node_ = std::make_shared<TesterNode>();
+    tester_node_->create_transform_buffer();
+    ndt_amcl_node_->set_parameter(rclcpp::Parameter("map_path", kTestMapPath));
+  }
+
+  void TearDown() override { rclcpp::shutdown(); }
+
+  bool wait_for_initialization() {
+    return spin_until([this] { return ndt_amcl_node_->is_initialized(); }, 200ms, ndt_amcl_node_, tester_node_);
+  }
+
+  bool wait_for_pose_estimate() {
+    tester_node_->latest_pose().reset();
+    tester_node_->create_pose_subscriber();
+    return spin_until([this] { return tester_node_->latest_pose().has_value(); }, 1000ms, ndt_amcl_node_, tester_node_);
+  }
+
+  bool wait_for_particle_cloud() {
+    tester_node_->create_particle_cloud_subscriber();
+    return spin_until(
+        [this] { return tester_node_->latest_particle_cloud().has_value(); }, 1000ms, ndt_amcl_node_, tester_node_);
+  }
+
+ protected:
+  std::shared_ptr<AmclNodeUnderTest> ndt_amcl_node_;
+  std::shared_ptr<TesterNode> tester_node_;
+};
+
```

Comment:
@serraramiro1 nit: consider relocating all test helpers to a separate header so we can reuse them for regular AMCL and NDT based AMCL tests.